timemachine/fe/single_topology.py

```diff
@@ -41,6 +42,45 @@
 OpenMMTopology = Any
 
 
+# Master Schedule
+# tbd: separate torsions into proper and improper later
+def _flip_min_max(min_max):
+    lamb_min, lamb_max = min_max
+    return 1 - lamb_max, 1 - lamb_min
+
+
+# core-bonds are never turned off if enforce_core_core=True
+CORE_BOND_MIN_MAX = [0.0, 1.0]
+CORE_ANGLE_MIN_MAX = [0.0, 1.0]
+CORE_TORSION_MIN_MAX = [0.0, 1.0]
+
+# core terms that are involved in chiral volumes being turned on or off
+CORE_CHIRAL_ATOM_CONVERTING_ON_MIN_MAX = [0.0, 0.5]
+CORE_CHIRAL_ANGLE_CONVERTING_ON_MIN_MAX = [0.5, 1.0]
+CORE_CHIRAL_ATOM_CONVERTING_OFF_MIN_MAX = _flip_min_max(CORE_CHIRAL_ATOM_CONVERTING_ON_MIN_MAX)
+CORE_CHIRAL_ANGLE_CONVERTING_OFF_MIN_MAX = _flip_min_max(CORE_CHIRAL_ANGLE_CONVERTING_ON_MIN_MAX)
+
+# non-converting (may be consistently in chirality or just achiral) dummy B groups that are turning on
+DUMMY_B_BOND_MIN_MAX = [0.0, 0.7]
+DUMMY_B_ANGLE_MIN_MAX = [0.0, 0.7]
+DUMMY_A_BOND_MIN_MAX = _flip_min_max(DUMMY_B_BOND_MIN_MAX)
+DUMMY_A_ANGLE_MIN_MAX = _flip_min_max(DUMMY_B_ANGLE_MIN_MAX)
+
+# chiral and converting dummy B groups are turning on
+DUMMY_B_CHIRAL_BOND_CONVERTING_ON_MIN_MAX = [0.0, 0.7]
+DUMMY_B_CHIRAL_ATOM_CONVERTING_ON_MIN_MAX = [0.3, 0.5]
+DUMMY_B_CHIRAL_ANGLE_CONVERTING_ON_MIN_MAX = [0.5, 0.7]  # angles are all turned off
+
+# chiral and converting dummy A groups are turning off
+DUMMY_A_CHIRAL_BOND_CONVERTING_OFF_MIN_MAX = _flip_min_max(DUMMY_B_CHIRAL_BOND_CONVERTING_ON_MIN_MAX)
+DUMMY_A_CHIRAL_ATOM_CONVERTING_OFF_MIN_MAX = _flip_min_max(DUMMY_B_CHIRAL_ATOM_CONVERTING_ON_MIN_MAX)
+DUMMY_A_CHIRAL_ANGLE_CONVERTING_OFF_MIN_MAX = _flip_min_max(DUMMY_B_CHIRAL_ANGLE_CONVERTING_ON_MIN_MAX)
+
+# torsions are the same throughout
+DUMMY_B_TORSION_MIN_MAX = [0.7, 1.0]  # chiral and achiral both use the same min/max for torsions
+DUMMY_A_TORSION_MIN_MAX = _flip_min_max(DUMMY_B_TORSION_MIN_MAX)
```

Comment:
Great point - hm let me think about this a bit. I debated something like this but I didn't want to give a false impression that the boundaries are "firm", i.e. there's probably some lee-way in how much they can overlap. (eg. there's already overlap between
`DUMMY_B_CHIRAL_BOND_CONVERTING_ON_MIN_MAX = [0.0, 0.7]`
and
`DUMMY_B_CHIRAL_ATOM_CONVERTING_ON_MIN_MAX = [0.3, 0.5]`

Even the left-bound of torsion checkpoint `DUMMY_B_TORSION_MIN_MAX = [0.7, 1.0]` can probably overlap with the right-bound of the angle checkpoint:

`DUMMY_B_CHIRAL_ANGLE_CONVERTING_ON_MIN_MAX = [0.5, 0.7]`

The key distinction here is that an angle term doesn't have to *fully* convert to its end-state before turning on the *torsion* term, it just has to have a reasonable force constant.

Reply:
> The key distinction here is that an angle term doesn't have to fully convert to its end-state before turning on the torsion term, it just has to have a reasonable force constant.

That's interesting.

Mostly, I'm unsure if there are 5 independent barrier-location parameters in this schedule that all happen to == `0.7` at the moment (but which _could_ all vary independently in the future), or if there are fewer independent parameters. Also, maybe some of the `0.7` barriers need to be coordinated with the `0.3` barriers...

If the ~10 literals (excluding `0.0, 1.0`) that appear in the schedule are all independent already, or if the amount of "slack" in each barrier is pretty large, maybe no change to code or documentation is warranted...

---------------

[tangent] Out of scope for current PR, but do you think we can make sharper assertions about _individual terms_ (e.g. "force constant on bond (5,6) must be greater than some threshold before the force constant on torsion (1,4,5,6) is > some threshold") rather than whole groups of terms ("activate all dummy bonds before any dummy torsions")?

If we can develop a reasonably complete list of such local assertion / constraint expressions (maybe the list will have a lot of special cases for ring-opening...), maybe we can then generate a fine-grained (i.e. per-term) force-constant-interpolation schedule for any given transformation, satisfying the constraints. @mcwitt and I discussed something along these lines a while ago -- maybe this could be useful to flesh out in a later PR.

Reply:
> If the ~10 literals (excluding 0.0, 1.0) that appear in the schedule are all independent already, or if the amount of "slack" in each barrier is pretty large, maybe no change to code or documentation is warranted...

I do think documentation is warranted re: slack at the boundaries. I will add it. 

> [tangent] Out of scope for current PR, but do you think we can make sharper assertions about individual terms (e.g. "force constant on bond (5,6) must be greater than some threshold before the force constant on torsion (1,4,5,6) is > some threshold") rather than whole groups of terms ("activate all dummy bonds before any dummy torsions")?

I think that's reasonable, but it will require a lot of careful testing on small systems. Roughly, there's two classes of assertions:

1) Assertions for numerical stability (eg. angle terms going to 180 degrees) can lead to large forces in torsions, etc. This class of assertions would be dependent on the time-step, whether or not we use HMR, etc. other numerical details.
2) Assertions for "trapped" sampling (eg. chiral restraints), these are dependent on length of simulation, number of expected flips in a typical 2ns sampling run, whether or not we use HREX/REST, etc.

I don't know really know how to address them systematically right now.

Reply:
For the case of chiral volumes, I think there's a simple assertion we can make. I can set that up as an example.

Reply:
Added an assertion `assert_bonds_defined_for_chiral_volumes` to check for bonds.

We can add some more assertions later, eg:

1) checking that "naked pyramids" without chiral volumes defined are invertible (by asserting that angles are close to 120, low force constants, ring membership, presence of impropers, etc.)
2) checking that impropers are not conflicting with highly pyramidal structures (right now proper + impropers are collected into a single type for interpolation purposes, so it's hard to separate).